# Case Study: Automated Machine Learning (AutoML) for Autonomous Intrusion Detection System Development 
This is the code for the paper entitled "**[Enabling AutoML for Zero-Touch Network Security: Use-Case Driven Analysis](https://ieeexplore.ieee.org/document/10472316)**" published in *IEEE Transactions on Network and Service Management* (IF:5.3).<br>
Authors: Li Yang (liyanghart@gmail.com), Mirna El Rajab, Abdallah Shami, and Sami Muhaidat<br>

L. Yang, M. E. Rajab, A. Shami, and S. Muhaidat, "Enabling AutoML for Zero-Touch Network Security: Use-Case Driven Analysis," IEEE Transactions on Network and Service Management, pp. 1-28, 2024, doi: https://doi.org/10.1109/TNSM.2024.3376631.

# Code Part 1: Automated Offline/Static/Batch Learning
Batch learning: Batch learning methods analyze static data in batches and often need access to the entire dataset prior to model training. Traditional ML algorithms can effectively solve batch learning tasks. Although batch learning models often achieve high performance due to their ability to learn diverse data patterns, it is often difficult to update these models once created. Therefore, batch learning faces two significant challenges: model degradation and data unavailability.

## Dataset 1: CICIDS2017
A subset of the network traffic data randomly sampled from the [CICIDS2017 dataset](https://www.unb.ca/cic/datasets/ids-2017.html).  

The Canadian Institute for Cybersecurity Intrusion Detection System 2017 (CICIDS2017) dataset has the most updated network threats. The CICIDS2017 dataset is close to real-world network data since it has a large amount of network traffic data, a variety of network features, various types of attacks, and highly imbalanced classes.

## Import libraries

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split,cross_val_score
import lightgbm as lgb
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score, precision_score, recall_score, f1_score
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.feature_selection import RFE
from scipy.stats import shapiro
from imblearn.over_sampling import SMOTE, ADASYN
from hyperopt import fmin, tpe, hp
import time

In [2]:
import warnings 
warnings.filterwarnings('ignore')

## Read the sampled CICIDS2017 dataset

In [3]:
df = pd.read_csv("Data/cic_0.01km.csv")

In [4]:
df

Flow Duration  Total Length of Fwd Packets  Fwd Packet Length Max  \
0              50833                            0                      0   
1                 49                            0                      0   
2                306                            6                      6   
3              63041                           65                     65   
4              47682                           43                     43   
...              ...                          ...                    ...   
28298             45                            0                      0   
28299      114309573                          511                    427   
28300          48850                           80                     40   
28301            260                           66                     33   
28302             25                            6                      6   

       Fwd Packet Length Mean  Bwd Packet Length Max  Bwd Packet Length Min  \
0                      0.0000                      0                      0   
1                      0.0000                      0                      0   
2                      6.0000                      6                      6   
3                     65.0000                    124                    124   
4                     43.0000                     59                     59   
...                       ...                    ...                    ...   
28298                  0.0000                      0                      0   
28299                 31.9375                    746                      0   
28300                 40.0000                     72                     72   
28301                 33.0000                     97                     97   
28302                  6.0000                      6                      6   

       Flow IAT Mean  Flow IAT Min  Fwd IAT Min  Fwd Header Length  \
0       5.083300e+04         50833            0                 32   
1       4.900000e+01            49           49                 64   
2       3.060000e+02           306            0                 20   
3       6.304100e+04         63041            0                 32   
4       4.768200e+04         47682            0                 32   
...              ...           ...          ...                ...   
28298   4.500000e+01            45            0                 32   
28299   3.941709e+06            94          165                332   
28300   1.628333e+04             1           48                 64   
28301   8.666667e+01            48           48                 40   
28302   2.500000e+01            25            0                 20   

       Bwd Header Length  Fwd Packets/s  Bwd Packets/s  Min Packet Length  \
0                     32      19.672260      19.672260                  0   
1                      0   40816.326530       0.000000                  0   
2                     20    3267.973856    3267.973856                  6   
3                     32      15.862693      15.862693                 65   
4                     32      20.972275      20.972275                 43   
...                  ...            ...            ...                ...   
28298                 32   22222.222220   22222.222220                  0   
28299                424       0.139971       0.122474                  0   
28300                 64      40.941658      40.941658                 40   
28301                 40    7692.307692    7692.307692                 33   
28302                 20   40000.000000   40000.000000                  6   

       URG Flag Count  Down/Up Ratio  Init_Win_bytes_forward  \
0                   1              1                     319   
1                   0              0                     277   
2                   0              1                       0   
3                   0              1                      -1   
4                   0              1                      -

# 1. Automated Data Pre-Processing

## Automated Transformation/Encoding
Automatically identify and transform string/text features into numerical features by the label encoding method to make the data more readable by ML models

In [5]:
# Define the automated data encoding function
def Auto_Encoding(df):
    cat_features=[x for x in df.columns if df[x].dtype=="object"] ## Find string/text features
    le=LabelEncoder()
    for col in cat_features:
        if col in df.columns:
            i = df.columns.get_loc(col)
            # Transform to numerical features
            df.iloc[:,i] = df.apply(lambda i:le.fit_transform(i.astype(str)), axis=0, result_type='expand')
    return df

In [6]:
df=Auto_Encoding(df)

## Automated Imputation
Detect and impute missing values to improve data quality

In [7]:
# Define the automated data imputation function
def Auto_Imputation(df):
    if df.isnull().values.any() or np.isinf(df).values.any(): # if there is any empty or infinite values
        df.replace([np.inf, -np.inf], np.nan, inplace=True)
        df.fillna(df.median(), inplace = True)  # Replace empty values with median values; there are other imputation methods discussed in the paper
    return df

In [8]:
df=Auto_Imputation(df)

## Automated normalization
Normalize the range of features to a similar scale to improve data quality

In [9]:
def Auto_Normalization(df):
    stat, p = shapiro(df)
    print('Statistics=%.3f, p=%.3f' % (stat, p))
    # interpret
    alpha = 0.05
    numeric_features = df.drop(['Label'],axis = 1).dtypes[df.dtypes != 'object'].index
    
    # check if the data distribution follows a Gaussian/normal distribution
    # If so, select the Z-score normalization method; otherwise, select the min-max normalization
    # Details are in the paper
    if p > alpha:
        print('Sample looks Gaussian (fail to reject H0)')
        df[numeric_features] = df[numeric_features].apply(
            lambda x: (x - x.mean()) / (x.std()))
        print('Z-score normalization is automatically chosen and used')
    else:
        print('Sample does not look Gaussian (reject H0)')
        df[numeric_features] = df[numeric_features].apply(
            lambda x: (x - x.min()) / (x.max()-x.min()))
        print('Min-max normalization is automatically chosen and used')
    return df

In [10]:
df=Auto_Normalization(df)

Statistics=0.076, p=0.000
Sample does not look Gaussian (reject H0)
Min-max normalization is automatically chosen and used


## Train-test split
Split the dataset into the training and the test set

In [11]:
X = df.drop(['Label'],axis=1)
y = df['Label']

# Here we used the 80%/20% split, it can be changed based on specific tasks
#X_train, X_test, y_train, y_test = train_test_split(X,y, train_size = 0.8, test_size = 0.2, shuffle=False,random_state = 0)
X_train, X_test, y_train, y_test = train_test_split(X,y, train_size = 0.8, test_size = 0.2,random_state = 0)

## Automated data balancing
Generate minority class samples to solve class-imbalance and improve data quality.  
Synthetic Minority Over-sampling Technique (SMOTE) method is used.

In [12]:
pd.Series(y_train).value_counts()

0    18126
1     4516
Name: Label, dtype: int64

In [13]:
# For binary data (can be modified for multi-class data with the same logic)
def Auto_Balancing(X_train, y_train):
    number0 = pd.Series(y_train).value_counts().iloc[0]
    number1 = pd.Series(y_train).value_counts().iloc[1]
    
    if number0 > number1:
        nlarge = number0
    else:
        nlarge = number1
    
    # evaluate whether the incoming dataset is imbalanced (the abnormal/normal ratio is smaller than a threshold (e.g., 50%)) 
    if (number1/number0 > 1.5) or (number0/number1 > 1.5):
        balanced=ADASYN(n_jobs=-1,sampling_strategy={0:nlarge, 1:nlarge})

        X_train, y_train = balanced.fit_resample(X_train, y_train)
        
    return X_train, y_train

In [14]:
X_train, y_train = Auto_Balancing(X_train, y_train)

In [15]:
pd.Series(y_train).value_counts()

0    18126
1    18102
Name: Label, dtype: int64

## Model learning (for Comparison)

In [16]:
%%time
lg = lgb.LGBMClassifier(verbose = -1)
lg.fit(X_train,y_train)
t1=time.time()
predictions = lg.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

Accuracy: 99.753%
Precision: 99.029%
Recall: 99.733%
F1-score: 99.38%
Time: 1.23332
Wall time: 191 ms


In [17]:
%%time
rf = RandomForestClassifier()
rf.fit(X_train,y_train)
t1=time.time()
predictions = rf.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

Accuracy: 99.823%
Precision: 99.38%
Recall: 99.733%
F1-score: 99.556%
Time: 7.75258
Wall time: 2.66 s


In [18]:
%%time
knn = KNeighborsClassifier()
knn.fit(X_train,y_train)
t1=time.time()
predictions = knn.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

Accuracy: 98.34%
Precision: 92.63900000000001%
Recall: 99.556%
F1-score: 95.973%
Time: 453.33682
Wall time: 2.58 s


In [19]:
import tensorflow as tf
from keras.layers import Input,Dense,Dropout,BatchNormalization,Activation
from keras import Model
import keras.backend as K
import keras.callbacks as kcallbacks
from keras import optimizers
from keras.optimizers import Adam

from sklearn.model_selection import GridSearchCV
from keras.wrappers.scikit_learn import KerasClassifier
from keras.callbacks import EarlyStopping
def ANN(optimizer = 'sgd',neurons=32,batch_size=1024,epochs=80,activation='relu',patience=8,loss='binary_crossentropy'):
    K.clear_session()
    inputs=Input(shape=(X.shape[1],))
    x=Dense(1000)(inputs)
    x=BatchNormalization()(x)
    x=Activation('relu')(x)
    x=Dropout(0.3)(x)
    x=Dense(256)(inputs)
    x=BatchNormalization()(x)
    x=Activation('relu')(x)
    x=Dropout(0.25)(x)
    x=Dense(2,activation='softmax')(x)
    model=Model(inputs=inputs,outputs=x,name='base_nlp')
    model.compile(optimizer='adam',loss='categorical_crossentropy')
#     model.compile(optimizer=Adam(lr = 0.01),loss='categorical_crossentropy',metrics=['accuracy'])
    early_stopping = EarlyStopping(monitor="loss", patience = patience)# early stop patience
    history = model.fit(X, pd.get_dummies(y).values,
              batch_size=batch_size,
              epochs=epochs,
              callbacks = [early_stopping],
              verbose=0) #verbose set to 1 will show the training process
    return model

Using TensorFlow backend.


In [21]:
%%time
ann = KerasClassifier(build_fn=ANN, verbose=0)
ann.fit(X_train,y_train)
predictions = ann.predict(X_test)
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

Accuracy: 88.801%
Precision: 64.199%
Recall: 98.667%
F1-score: 77.786%
Time: 453.33682
Wall time: 7.98 s


# 2. Automated Feature Engineering
Feature selection method 1: **Recursive Feature Elimination (RFE)**, used to remove irrelevant features to improve model efficiency  
Feature selection method 2: **Pearson Correlation**, used to remove redundant features to improve model efficiency and accuracy  

In [29]:
def Feature_Importance_RFE(data, n_features_to_select=20):
    features = data.drop(['Label'], axis=1).values  # "Label" should be changed to the target class variable name if different
    labels = data['Label'].values

    # Extract feature names
    feature_names = list(data.drop(['Label'], axis=1).columns)

    # Create a base estimator
    model = lgb.LGBMRegressor(verbose = -1)

    # Create the RFE object and rank each feature
    rfe = RFE(estimator=model, n_features_to_select=n_features_to_select, step=1)
    rfe.fit(features, labels)

    # Get the feature ranking
    feature_ranking = rfe.ranking_

    # Create a DataFrame for feature importances
    feature_importances = pd.DataFrame({'feature': feature_names, 'ranking': feature_ranking})

    # Sort features according to their ranking
    feature_importances = feature_importances.sort_values('ranking', ascending=True).reset_index(drop=True)

    # Get the features to drop
    to_drop = list(feature_importances[feature_importances['ranking'] > 1]['feature'])

    return to_drop

In [30]:
# Remove redundant features
def Feature_Redundancy_Pearson(data):
    correlation_threshold=0.90 # Only remove features with the redundancy>90%. It can be changed
    features = data.drop(['Label'],axis=1)
    corr_matrix = features.corr()

    # Extract the upper triangle of the correlation matrix
    upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k = 1).astype(np.bool))

    # Select the features with correlations above the threshold
    # Need to use the absolute value
    to_drop = [column for column in upper.columns if any(upper[column].abs() > correlation_threshold)]

    # Dataframe to hold correlated pairs
    record_collinear = pd.DataFrame(columns = ['drop_feature', 'corr_feature', 'corr_value'])

    # Iterate through the columns to drop
    for column in to_drop:

        # Find the correlated features
        corr_features = list(upper.index[upper[column].abs() > correlation_threshold])

        # Find the correlated values
        corr_values = list(upper[column][upper[column].abs() > correlation_threshold])
        drop_features = [column for _ in range(len(corr_features))]    

        # Record the information (need a temp df for now)
        temp_df = pd.DataFrame.from_dict({'drop_feature': drop_features,
                                         'corr_feature': corr_features,
                                         'corr_value': corr_values})
        record_collinear = record_collinear.append(temp_df, ignore_index = True)
#     print(record_collinear)
    return to_drop

In [31]:
def Auto_Feature_Engineering(df):
    drop1 = Feature_Importance_RFE(df)
    dfh1 = df.drop(columns = drop1)
    
    drop2 = Feature_Redundancy_Pearson(dfh1)
    dfh2 = dfh1.drop(columns = drop2)
    
    return dfh2

In [32]:
dfh2 = Auto_Feature_Engineering(df)
dfh2

Flow Duration  Total Length of Fwd Packets  Fwd Packet Length Max  \
0       4.236419e-04                     0.000000               0.000000   
1       4.416669e-07                     0.000000               0.000000   
2       2.583334e-06                     0.000008               0.000242   
3       5.253752e-04                     0.000090               0.002619   
4       3.973835e-04                     0.000060               0.001732   
...              ...                          ...                    ...   
28298   4.083335e-07                     0.000000               0.000000   
28299   9.525802e-01                     0.000710               0.017204   
28300   4.071168e-04                     0.000111               0.001612   
28301   2.200001e-06                     0.000092               0.001330   
28302   2.416668e-07                     0.000008               0.000242   

       Fwd Packet Length Mean  Bwd Packet Length Max  Bwd Packet Length Min  \
0                    0.000000               0.000000               0.000000   
1                    0.000000               0.000000               0.000000   
2                    0.001556               0.000516               0.004144   
3                    0.016856               0.010660               0.085635   
4                    0.011151               0.005072               0.040746   
...                       ...                    ...                    ...   
28298                0.000000               0.000000               0.000000   
28299                0.008282               0.064133               0.000000   
28300                0.010373               0.006190               0.049724   
28301                0.008558               0.008339               0.066989   
28302                0.001556               0.000516               0.004144   

       Flow IAT Mean  Flow IAT Min   Fwd IAT Min  Fwd Header Length  \
0       4.707129e-04  4.707129e-04  0.000000e+00           0.000150   
1       4.907407e-07  4.907407e-07  4.083333e-07           0.000299   
2       2.870370e-06  2.870370e-06  0.000000e+00           0.000094   
3       5.837500e-04  5.837500e-04  0.000000e+00           0.000150   
4       4.415370e-04  4.415370e-04  0.000000e+00           0.000150   
...              ...           ...           ...                ...   
28298   4.537037e-07  4.537037e-07  0.000000e+00           0.000150   
28299   3.649735e-02  9.074074e-07  1.375000e-06           0.001554   
28300   1.508086e-04  4.629629e-08  4.000000e-07           0.000299   
28301   8.395061e-07  4.814815e-07  4.000000e-07           0.000187   
28302   2.685185e-07  2.685185e-07  0.000000e+00           0.000094   

       Fwd Packets/s  Bwd Packets/s  Min Packet Length  URG Flag Count  \
0       6.557420e-06   1.311484e-05           0.000000             1.0   
1       1.360544e-02   0.000000e+00           0.000000             0.0   
2       1.089325e-03   2.178649e-03           0.017804             0.0   
3       5.287564e-06   1.057513e-05           0.192878             0.0   
4       6.990758e-06   1.398152e-05           0.127596             0.0   
...              ...            ...                ...             ...   
28298   7.407407e-03   1.481481e-02           0.000000             1.0   
28299   4.665693e-08   8.164962e-08           0.000000             0.0   
28300   1.364722e-05   2.729444e-05           0.118694             0.0   
28301   2.564103e-03   5.128205e-03           0.097923             0.0   
28302   1.333333e-02   2.666667e-02           0.017804             1.0   

       Down/Up Ratio  Init_Win_bytes_forward  Init_Win_bytes_backward  \
0           0.142857                0.004883                 0.002350   
1           0.000000                0.004242                 0.000000   
2           0.142857                0.000015                 0.000015   
3           0.142857                0.000000                 0.000000   
4           0.142857                0.

## Data Split & Balancing (After Feature Engineering)

In [33]:
X = dfh2.drop(['Label'],axis=1)
y = dfh2['Label']

#X_train, X_test, y_train, y_test = train_test_split(X,y, train_size = 0.8, test_size = 0.2, shuffle=False,random_state = 0)
X_train, X_test, y_train, y_test = train_test_split(X,y, train_size = 0.8, test_size = 0.2,random_state = 0)

In [34]:
X_train, y_train = Auto_Balancing(X_train, y_train)

# 3. Automated Model Selection
Select the best-performing model among five common machine learning models (Naive Bayes, KNN, random forest, LightGBM, and ANN/MLP) by evaluating their learning performance

### Grid Search

In [35]:
# Create a pipeline
pipe = Pipeline([('classifier', KNeighborsClassifier())])

# Create space of candidate learning algorithms and their hyperparameters
search_space = [
                {'classifier': [KNeighborsClassifier()]},
                {'classifier': [RandomForestClassifier()]},
                {'classifier': [lgb.LGBMClassifier(verbose = -1)]},
                {'classifier': [KerasClassifier(build_fn=ANN, verbose=0)]},
                 ]

In [36]:
clf = GridSearchCV(pipe, search_space, cv=3, verbose=0)

In [37]:
clf.fit(X, y)

GridSearchCV(cv=3,
             estimator=Pipeline(steps=[('classifier', KNeighborsClassifier())]),
             param_grid=[{'classifier': [KNeighborsClassifier()]},
                         {'classifier': [RandomForestClassifier()]},
                         {'classifier': [LGBMClassifier(verbose=-1)]},
                         {'classifier': [<keras.wrappers.scikit_learn.KerasClassifier object at 0x000001D50A392988>]}])

In [38]:
print("Best Model:"+ str(clf.best_params_))
print("Accuracy:"+ str(clf.best_score_))

Best Model:{'classifier': LGBMClassifier(verbose=-1)}
Accuracy:0.952513976271599


In [39]:
clf.cv_results_

{'mean_fit_time': array([3.98953756e-03, 1.04863024e+00, 1.10047976e-01, 7.77408489e+00]),
 'std_fit_time': array([0.00081391, 0.05885326, 0.00497592, 0.3468082 ]),
 'mean_score_time': array([2.47092915, 0.05418356, 0.00930921, 0.2336181 ]),
 'std_score_time': array([0.03166898, 0.00168619, 0.00047042, 0.03251903]),
 'param_classifier': masked_array(data=[KNeighborsClassifier(), RandomForestClassifier(),
                    LGBMClassifier(verbose=-1),
              mask=[False, False, False, False],
        fill_value='?',
             dtype=object),
 'params': [{'classifier': KNeighborsClassifier()},
  {'classifier': RandomForestClassifier()},
  {'classifier': LGBMClassifier(verbose=-1)},
  {'classifier': <keras.wrappers.scikit_learn.KerasClassifier at 0x1d50a392988>}],
 'split0_test_score': array([0.881823  , 0.90768415, 0.94944356,        nan]),
 'split1_test_score': array([0.91912232, 0.92993428, 0.91763833,        nan]),
 'split2_test_score': array([0.96979012, 0.98367607, 0.99046

LightGBM model is the best performing machine learning model, and the best cross-validation accuracy is 95.251%

## Model learning (for Comparison)

In [40]:
%%time
lg = lgb.LGBMClassifier(verbose = -1)
lg.fit(X_train,y_train)
t1=time.time()
predictions = lg.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

Accuracy: 99.717%
Precision: 98.855%
Recall: 99.733%
F1-score: 99.292%
Time: 0.8809
Wall time: 184 ms


In [41]:
%%time
rf = RandomForestClassifier()
rf.fit(X_train,y_train)
t1=time.time()
predictions = rf.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

Accuracy: 99.84100000000001%
Precision: 99.468%
Recall: 99.733%
F1-score: 99.601%
Time: 7.75292
Wall time: 2.72 s


In [42]:
%%time
knn = KNeighborsClassifier()
knn.fit(X_train,y_train)
t1=time.time()
predictions = knn.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

Accuracy: 98.21600000000001%
Precision: 92.036%
Recall: 99.644%
F1-score: 95.68900000000001%
Time: 468.84165
Wall time: 2.66 s


In [43]:
import tensorflow as tf
from keras.layers import Input,Dense,Dropout,BatchNormalization,Activation
from keras import Model
import keras.backend as K
import keras.callbacks as kcallbacks
from keras import optimizers
from keras.optimizers import Adam

from sklearn.model_selection import GridSearchCV
from keras.wrappers.scikit_learn import KerasClassifier
from keras.callbacks import EarlyStopping
def ANN(optimizer = 'sgd',neurons=32,batch_size=1024,epochs=80,activation='relu',patience=8,loss='binary_crossentropy'):
    K.clear_session()
    inputs=Input(shape=(X_train.shape[1],))
    x=Dense(1000)(inputs)
    x=BatchNormalization()(x)
    x=Activation('relu')(x)
    x=Dropout(0.3)(x)
    x=Dense(256)(inputs)
    x=BatchNormalization()(x)
    x=Activation('relu')(x)
    x=Dropout(0.25)(x)
    x=Dense(2,activation='softmax')(x)
    model=Model(inputs=inputs,outputs=x,name='base_nlp')
    model.compile(optimizer='adam',loss='categorical_crossentropy')
#     model.compile(optimizer=Adam(lr = 0.01),loss='categorical_crossentropy',metrics=['accuracy'])
    early_stopping = EarlyStopping(monitor="loss", patience = patience)# early stop patience
    history = model.fit(X_train, pd.get_dummies(y_train).values,
              batch_size=batch_size,
              epochs=epochs,
              callbacks = [early_stopping],
              verbose=0) #verbose set to 1 will show the training process
    return model

In [44]:
%%time
ann = KerasClassifier(build_fn=ANN, verbose=0)
ann.fit(X_train,y_train)
predictions = ann.predict(X_test)
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

Accuracy: 88.39399999999999%
Precision: 63.31099999999999%
Recall: 98.933%
F1-score: 77.211%
Time: 468.84165
Wall time: 10.8 s


# 4. Hyperparameter Optimization
Optimize the best performing machine learning model (lightGBM) by tuning its hyperparameters

## Hold-out validation

In [64]:
#Particle Swarm Optimization
import optunity
import optunity.metrics

# Define the hyperparameter configuration space
search = {
    'n_estimators': [50, 500],
    'max_depth': [5, 50],
    'learning_rate': (0, 1),
    "num_leaves":[100, 2000],
    "min_child_samples":[10, 50],
         }
# Define the objective function
def performance(n_estimators=None, max_depth=None,learning_rate=None,num_leaves=None,min_child_samples=None):
    clf = lgb.LGBMClassifier(n_estimators=int(n_estimators),
                                   max_depth=int(max_depth),
                                   learning_rate=float(learning_rate),
                                   num_leaves=int(num_leaves),
                                   min_child_samples=int(min_child_samples),
                                  )
    clf.fit(X_train,y_train)
    prediction = clf.predict(X_test)
    score = accuracy_score(y_test,prediction)
    return score

# Detect the optimal hyperparameter values
optimal_configuration, info, _ = optunity.maximize(performance,
                                                  solver_name='particle swarm',
                                                  num_evals=20,
                                                   **search
                                                  )
print(optimal_configuration)
print("Accuracy:"+ str(info.optimum))

{'n_estimators': 419.580078125, 'max_depth': 43.4521484375, 'learning_rate': 0.2587890625, 'num_leaves': 1163.18359375, 'min_child_samples': 38.7109375}
Accuracy:0.9980568804098215


In [68]:
%%time
clf = lgb.LGBMClassifier(max_depth=43, learning_rate=  0.2587890625, n_estimators = 419, 
                         num_leaves = 1163, min_child_samples = 38)
clf.fit(X_train,y_train)
predictions = clf.predict(X_test)
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")

Accuracy: 99.806%
Precision: 99.37899999999999%
Recall: 99.644%
F1-score: 99.512%
Wall time: 1.66 s


After hyperparameter optimization, the hold-out accuracy has been improved from 99.717% to 99.806%

In [69]:
import optunity
import optunity.metrics
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Define the hyperparameter configuration space for RandomForestClassifier
search = {
    'n_estimators': [50, 500],  # Number of trees in the forest
    'max_depth': [5, 50],  # Maximum depth of the tree
    'min_samples_split': [2, 11],  # Minimum number of samples required to split an internal node
    'min_samples_leaf': [1, 11],  # Minimum number of samples required to be at a leaf node
    'criterion_index': [0, 1],  # Index to select criterion, 0 for "gini", 1 for "entropy"
}

# Define the objective function for RandomForestClassifier
def performance(n_estimators=None, max_depth=None, min_samples_split=None, min_samples_leaf=None, criterion_index=None):
    # Convert criterion_index to actual criterion string
    criterion = ["gini", "entropy"][int(criterion_index)]
    
    # Define and fit the model
    clf = RandomForestClassifier(
        n_estimators=int(n_estimators),
        max_depth=int(max_depth),
        min_samples_split=int(min_samples_split),
        min_samples_leaf=int(min_samples_leaf),
        criterion=criterion
    )
    clf.fit(X_train, y_train)
    prediction = clf.predict(X_test)
    score = accuracy_score(y_test, prediction)
    return score

# Detect the optimal hyperparameter values using PSO
optimal_configuration, info, _ = optunity.maximize(performance,
                                                   solver_name='particle swarm',
                                                   num_evals=20,
                                                   **search
                                                  )

print(optimal_configuration)
print("Accuracy:" + str(info.optimum))


{'n_estimators': 113.5009765625, 'max_depth': 40.96923828125, 'min_samples_split': 6.03857421875, 'min_samples_leaf': 1.8935546875, 'criterion_index': 0.30224609375}
Accuracy:0.9982335276452924


In [72]:
%%time
clf = RandomForestClassifier(max_depth=40, n_estimators = 113, min_samples_split = 6,
                         min_samples_leaf = 1, criterion = 'gini')
clf.fit(X_train,y_train)
predictions = clf.predict(X_test)
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")

Accuracy: 99.823%
Precision: 99.38%
Recall: 99.733%
F1-score: 99.556%
Wall time: 3.14 s


# 5. Combined Algorithm Selection and Hyperparameter tuning (CASH)
CASH is the process of combining the two AutoML procedures: model selection and hyperparameter optimization.

## Method: Particle Swarm Optimization (PSO)

In [73]:
import optunity
import optunity.metrics

search = {'algorithm': {'k-nn': {'n_neighbors': [3, 10]},
                        'random-forest': {
                                'n_estimators': [50, 500],
                                'max_features': [5, 12],
                                'max_depth': [5,50],
                                "min_samples_split":[2,11],
                                "min_samples_leaf":[1,11]},
                        'lightgbm': {
                                'n_estimators': [50, 500],
                                'max_depth': [5, 50],
                                'learning_rate': (0, 1),
                                "num_leaves":[100, 2000],
                                "min_child_samples":[10, 50],
                                    },
                        'ann': {
                                'neurons': [10, 100],
                                'epochs': [20, 50],
                                'patience': [3, 20],
                                }
                        }
          
         }
def performance(
                algorithm, n_neighbors=None, 
    n_estimators=None, max_features=None,max_depth=None,min_samples_split=None,min_samples_leaf=None,
    learning_rate=None,num_leaves=None,min_child_samples=None,
    neurons=None,epochs=None,patience=None
):
    # fit the model
    if algorithm == 'k-nn':
        model = KNeighborsClassifier(n_neighbors=int(n_neighbors))
    elif algorithm == 'random-forest':
        model = RandomForestClassifier(n_estimators=int(n_estimators),
                                       max_features=int(max_features),
                                       max_depth=int(max_depth),
                                       min_samples_split=int(min_samples_split),
                                       min_samples_leaf=int(min_samples_leaf))
    elif algorithm == 'lightgbm':
        model = lgb.LGBMClassifier(n_estimators=int(n_estimators),
                                   max_depth=int(max_depth),
                                   learning_rate=float(learning_rate),
                                   num_leaves=int(num_leaves),
                                   min_child_samples=int(min_child_samples),
                                  )
    elif algorithm == 'ann':
        model = KerasClassifier(build_fn=ANN, verbose=0,
                               neurons=int(neurons),
                                epochs=int(epochs),
                                patience=int(patience)
                               )
    else:
        raise ArgumentError('Unknown algorithm: %s' % algorithm)
# predict the test set
    model.fit(X_train,y_train)
    prediction = model.predict(X_test)
    score = accuracy_score(y_test,prediction)
    return score

# Run the CASH process
optimal_configuration, info, _ = optunity.maximize_structured(performance, 
                                                              search_space=search, 
                                                              num_evals=10)
print(optimal_configuration)
print(info.optimum)

{'algorithm': 'random-forest', 'epochs': None, 'neurons': None, 'patience': None, 'n_neighbors': None, 'learning_rate': None, 'max_depth': 31.3232421875, 'min_child_samples': None, 'n_estimators': 191.064453125, 'num_leaves': None, 'max_features': 5.4443359375, 'min_samples_leaf': 1.185546875, 'min_samples_split': 8.1787109375}
0.9982335276452924


In [74]:
%%time
clf = RandomForestClassifier(max_depth=31, n_estimators = 191, max_features=5, min_samples_split = 8,
                         min_samples_leaf = 1, criterion = 'gini')
clf.fit(X_train,y_train)
predictions = clf.predict(X_test)
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")

Accuracy: 99.823%
Precision: 99.468%
Recall: 99.644%
F1-score: 99.556%
Wall time: 6.76 s


Random Forest with the above hyperparameter values is identified as the optimal model